{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "51d7723d-b35e-42f0-b220-dc8bc4ba820d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import subprocess\n",
    "from glob import glob\n",
    "from datetime import datetime\n",
    "import numpy as np\n",
    "import re\n",
    "from glob import glob\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy.stats import pearsonr\n",
    "import matplotlib.pyplot as plt \n",
    "import gzip\n",
    "import dask.dataframe as dd\n",
    "from tqdm import tqdm\n",
    "import gc\n",
    "from collections import Counter\n",
    "from sklearn.decomposition import PCA\n",
    "from umap import UMAP\n",
    "from pathlib import Path\n",
    "import os\n",
    "import inspect\n",
    "from time import sleep\n",
    "import sys\n",
    "import itertools\n",
    "from IPython.utils import io\n",
    "import psycopg2\n",
    "import warnings\n",
    "import requests\n",
    "from io import StringIO\n",
    "from tqdm import tqdm_notebook\n",
    "warnings.filterwarnings('ignore')\n",
    "#conda create --name gpipe -c conda-forge openjdk=17 ipykernel pandas seaborn scikit-learn umap-learn psycopg2 dask\n",
    "#conda activate gpipe\n",
    "#conda install -c bioconda gcta plink snpeff\n",
    "#wget https://snpeff.blob.core.windows.net/versions/snpEff_latest_core.zip\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "df430dc2-9d7f-4f90-a3c7-f419d1b769a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "class vcfGeneratorExitools:\n",
    "    def corrfunc(x, y, ax=None, **kws):\n",
    "        r, _ = pearsonr(x, y)\n",
    "        ax = ax or plt.gca()\n",
    "        ax.annotate(f'ρ = {r:.2f}', xy=(.1, .9), xycoords=ax.transAxes)\n",
    "\n",
    "    def get_vcf_header(vcf_path):\n",
    "        with gzip.open(vcf_path, \"rt\") as ifile:\n",
    "            for num, line in enumerate(ifile):\n",
    "                if line.startswith(\"#CHROM\"): return line.strip().split('\\t')\n",
    "                if num > 10000: return '-1'\n",
    "        return '-1'\n",
    "\n",
    "    def read_vcf(filename, method = 'pandas'):\n",
    "        if method == 'dask':\n",
    "            return dd.read_csv(filename,  compression='gzip', comment='#',  delim_whitespace=True, header=None, \n",
    "                               names = vcfGeneratorExitools.get_vcf_header(filename),blocksize=None,  dtype=str, ).repartition(npartitions = 100000)\n",
    "        # usecols=['#CHROM', 'POS']\n",
    "        return pd.read_csv(filename,  compression='gzip', comment='#',  delim_whitespace=True,\n",
    "                           header=None, names = vcfGeneratorExitools.get_vcf_header(filename),  dtype=str )\n",
    "\n",
    "    def name_gen2(filename):\n",
    "        return filename.split('/')[-1].split('.')[0]\n",
    "    \n",
    "    def get_vcf_metadata(vcf_path):\n",
    "        out = ''\n",
    "        with gzip.open(vcf_path, \"rt\") as ifile:\n",
    "            for num, line in enumerate(ifile):\n",
    "                if line.startswith(\"#CHROM\"): return out\n",
    "                out += line \n",
    "        return '-1'\n",
    "\n",
    "    def pandas2vcf(df, filename, metadata = ''):\n",
    "        if  metadata == '':\n",
    "            header = '\\n'.join([\"##fileformat=VCFv4.1\",\n",
    "            '##fileDate=20090805',\n",
    "            '##source=myImputationProgramV3.1',\n",
    "            '##reference=000GenomesPilot-NCBI36',\n",
    "            '##INFO=<ID=NS,Number=1,Type=Integer,Description=\"Number of Samples With Data\">',\n",
    "            '##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">',\n",
    "            '##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">',\n",
    "            '##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Read Depth\">',\n",
    "            '##FORMAT=<ID=HQ,Number=2,Type=Integer,Description=\"Haplotype Quality\">']) + '\\n'\n",
    "        elif metadata[-4:] == '.vcf': header = get_vcf_metadata(metadata)\n",
    "\n",
    "        with open(filename, 'w') as vcf: \n",
    "            vcf.write(header)\n",
    "        df.to_csv(filename, sep=\"\\t\", mode='a', index=False)\n",
    "\n",
    "    \n",
    "\n",
    "def bash(call, verbose = 0, return_stdout = True, print_call = True):\n",
    "    if print_call: print(call+'\\n')\n",
    "    out = subprocess.run(call.split(' '), capture_output = True) \n",
    "    if verbose and not return_stdout: print(out.stdout)\n",
    "    \n",
    "    if out.stderr: print(out.stderr.decode('ascii'))\n",
    "    if return_stdout: return out.stdout.decode('ascii').strip().split('\\n')\n",
    "    return out\n",
    "\n",
    "def qsub(call: str, queue = 'condo', walltime = 3, ppn = 1, out = 'log/', err = 'logerr/', project_dir = ''):\n",
    "    err_path = f'{project_dir}{err}$PBS_JOBNAME.err'\n",
    "    out_path = f'{project_dir}{out}$PBS_JOBNAME.out'\n",
    "    call_path = f'{project_dir}{call}'\n",
    "    return bash(f'qsub -q {queue} -l nodes=1:ppn={ppn} -j oe -o {out_path} -e {err_path} -l walltime={walltime}:00:00 {call_path}')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "23585ebb-c177-42a1-a8af-de81e8289b6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "class gwas_pipe:\n",
    "    def __init__(self, \n",
    "                 path: str = f'{Path().absolute()}/', \n",
    "                 use_tscc_modules: list = [],\n",
    "                 all_genotypes: str = '/projects/ps-palmer/apurva/riptide/genotypes/round9_1',\n",
    "                 gtca_path: str = '',\n",
    "                 data: pd.DataFrame() = pd.DataFrame(),\n",
    "                 traits: list = [],\n",
    "                 project_name: str = 'test',\n",
    "                 phewas_db: str = 'phewasdb.parquet.gz',\n",
    "                 threads: int = 12):\n",
    "        \n",
    "        '''\n",
    "        to run the gwas pipeline we need to ...\n",
    "        \n",
    "        '''\n",
    "        \n",
    "        if use_tscc_modules: bash(f'module load {\" \".join(use_tscc_modules)}')\n",
    "        self.gtca = 'gcta64' if not gtca_path else gtca_path\n",
    "        self.path = path\n",
    "        self.all_genotypes = all_genotypes\n",
    "        df = data\n",
    "        df.columns = df.columns.str.lower()\n",
    "        self.df = df\n",
    "        self.traits = [x.lower() for x in traits]\n",
    "        self.phewas_db = phewas_db\n",
    "        self.project_name = project_name\n",
    "        \n",
    "        self.sample_path = f'{self.path}genotypes/sample_rfids.txt'\n",
    "        self.genotypes_subset = f'{self.path}genotypes/genotypes'\n",
    "        self.genotypes_subset_vcf = f'{self.path}genotypes/genotypes_subset_vcf.vcf.gz'\n",
    "        \n",
    "        self.autoGRM = f'{self.path}grm/AllchrGRM'\n",
    "        self.xGRM = f'{path}grm/xchrGRM'\n",
    "        self.log = pd.DataFrame( columns = ['function', 'call', 'out'])\n",
    "        self.thrflag = f'--thread-num {threads}'\n",
    "        self.print_call = True\n",
    "        self.chrList = ['x' if x == 21 else x for x in range(1,22)]\n",
    "        \n",
    "    def plink2Df(self, call, temp_out_filename = 'temp/temp', dtype = 'ld'):\n",
    "        '''\n",
    "        assigned THIAGO\n",
    "        this function receives a plink call as a string, \n",
    "        runs the call, \n",
    "        then reads the output file with the ending dtype \n",
    "        and returns it as a pandas table\n",
    "        '''\n",
    "        \n",
    "        full_call = re.sub(r' +', ' ', call + f'--out {self.path}{temp_out_filename}')\n",
    "        \n",
    "        ### add line to delete temp_out_filename before doing the \n",
    "        bash(full_call, print_call = False)\n",
    "\n",
    "        try: out = pd.read_csv(f'{self.path}{temp_out_filename}.{dtype}', sep = '\\s+')\n",
    "        except:\n",
    "            print(f\"file not found\")\n",
    "            out = pd.DataFrame()\n",
    "        return out \n",
    "    \n",
    "    def plink(self, return_file = 'ld', outfile = '',  **kwargs):\n",
    "        '''\n",
    "        assigned THIAGO\n",
    "        this function is a wrapper to run plink as a python function\n",
    "        instead of having to have it as a string and bash call it\n",
    "        if writing a flag that doesn't require a variable e.g.\n",
    "        --make-grm use make_grm = ''\n",
    "        if outfile is empty and return file is not\n",
    "            we will return a pandas dataframe from the plink outputfile with ending return_file\n",
    "        otherwise we will save the files under the filename outfile\n",
    "        '''\n",
    "        \n",
    "        call = 'plink ' + ' '.join([f'--{k.replace(\"_\", \"-\")} {v}'\n",
    "                                    for k,v in kwargs.items()])\n",
    "        if not outfile and return_file:\n",
    "            return self.plink2Df(call ,dtype = f'{return_file}' ) \n",
    "        else:\n",
    "            bash(call + f' --out {outfile}', print_call=False)\n",
    "        return\n",
    "        \n",
    "    def bashLog(self, call, func, print_call = True):\n",
    "        self.append2log(func, call , bash(re.sub(r' +', ' ', call), print_call = print_call))\n",
    "        \n",
    "        \n",
    "    def append2log(self, func, call, out):\n",
    "        self.log.loc[len(self.log)] = [func, call, out]\n",
    "        with open(f'{self.path}/log/{func}.log', 'w') as f:\n",
    "            f.write('\\n'.join(out))\n",
    "            \n",
    "    def make_dir_structure(self, folders: list = ['data', 'genotypes', 'grm', 'log', 'logerr', \n",
    "                                            'results', 'temp', 'data/pheno', 'results/heritability', \n",
    "                                             'results/gwas',  'results/loco', 'results/qtls','results/eqtl',\n",
    "                                                  'results/phewas', 'temp/r2']):\n",
    "        \n",
    "        '''\n",
    "        assigned THIAGO\n",
    "        this function will create the folder structure for the outputfiles of the project \n",
    "        \n",
    "        '''\n",
    "        for folder in folders:\n",
    "            os.makedirs(f'{self.path}{folder}', exist_ok = True)\n",
    "            \n",
    "            \n",
    "    def subsetSamplesFromAllGenotypes(self,samplelist: list = [], \n",
    "                                      use_rfid_from_df = True, sourceFormat = 'vcf', \n",
    "                                      geno: float = .1, maf: float = .005, hwe: float = 1e-10 ):\n",
    "        \n",
    "        '''\n",
    "        this function will get the large round vcf (or plink) file, subset it based of the rfid from the dataframe\n",
    "        (or a sample list if wanted), and filter based on missingness, minor allele frequency and hwe equilibrium\n",
    "        '''\n",
    "        \n",
    "        \n",
    "        funcName = inspect.getframeinfo(inspect.currentframe()).function\n",
    "        \n",
    "        os.makedirs(f'{self.path}genotypes', exist_ok = True)\n",
    "        \n",
    "        df = self.df[['rfid', 'rfid']] if use_rfid_from_df else pd.DataFrame([samplelist,samplelist]).T.astype(str)\n",
    "        df.to_csv(f'{self.path}genotypes/sample_rfids.txt', index = False, header = None, sep = ' ')\n",
    "        self.sample_names = samplelist\n",
    "        #### add here a way so plink know which samples are male and which are female\n",
    "        \n",
    "        fmt_call = {'vcf': 'vcf', 'plink': 'bfile'}[sourceFormat]        \n",
    "        \n",
    "        extra_params = f'--geno {geno} --maf {maf} --hwe {hwe} --double-id --set-missing-var-ids @:#'\n",
    "        \n",
    "        self.bashLog(f'plink --{fmt_call} {self.all_genotypes} --keep {self.sample_path} {extra_params} {self.thrflag} --make-bed --out {self.genotypes_subset}',\n",
    "                    funcName)\n",
    "        \n",
    "    def SubsampleMissMafHweFilter(self, sexfmt: str = 'M|F',  sexColumn: str = 'sex',  \n",
    "                                  geno: float = .1, maf: float = .005, hwe: float = 1e-10,\n",
    "                                  sourceFormat = 'vcf', remove_dup: bool = True, print_call: bool = False):\n",
    "        \n",
    "        '''\n",
    "        more complex version of the function above that accounts for sample sex when subseting and performing the filters\n",
    "        i.e. hwe is more complicated in chr X, chr Y and MT \n",
    "        '''\n",
    "        \n",
    "        funcName = inspect.getframeinfo(inspect.currentframe()).function\n",
    "        \n",
    "               \n",
    "        fmt_call = {'vcf': 'vcf', 'plink': 'bfile'}[sourceFormat]  \n",
    "        rmv = f' --double-id --set-missing-var-ids @:# ' if remove_dup else ' '\n",
    "        sub = self.genotypes_subset\n",
    "        \n",
    "        mfList = sorted(sexfmt.split('|'))\n",
    "        self.samples = {}\n",
    "        for num, sx in enumerate(tqdm(mfList)): \n",
    "            \n",
    "            dff = self.df[(self.df[sexColumn] == sx) & \n",
    "                          (self.df.astype(str).rfid.isin(vcftools.get_vcf_header(self.all_genotypes)))]\n",
    "            dff[['rfid', 'rfid']].to_csv(f'{self.path}genotypes/sample_rfids_{sx}.txt', index = False, header = None, sep = ' ')\n",
    "            self.samples[sx] = f'{self.path}genotypes/sample_rfids_{sx}.txt'\n",
    "            \n",
    "            filtering_flags = f' --geno {geno} --maf {maf} --hwe {hwe}'\n",
    "            filtering_flags_justx = f''\n",
    "            extra_flags = f'--not-chr X'  if num == 1 else ''\n",
    "            \n",
    "            self.bashLog(f'plink --{fmt_call} {self.all_genotypes} --keep {self.samples[sx]} {filtering_flags} {rmv} {extra_flags} {self.thrflag} --make-bed --out {sub}_{sx}',\n",
    "                        f'{funcName}_subseting_{sx}')\n",
    "            \n",
    "            if num == 1:\n",
    "                self.bashLog(f'plink --bfile {sub}_{sx} --chr x {self.thrflag} --make-bed --out {sub}_{sx}_xchr',\n",
    "                        f'{funcName}_maleXsubset{sx}') #--out {sub}_{sx}_xchr\n",
    "                male_1_x_filenames = [aa for aa in [f'{sub}_{sx}', f'{sub}_{sx}_xchr'] if len(glob(aa+'.*')) >= 5]\n",
    "                male_gen_filenames = f'{self.path}/genotypes/temp_male_filenames'\n",
    "                pd.DataFrame(male_1_x_filenames).to_csv(male_gen_filenames, index = False, header = None)\n",
    "            else: female_hwe = f'{sub}_{sx}'\n",
    "                \n",
    "        print('merging sexes')        \n",
    "        self.bashLog(f'plink --bfile {female_hwe} --merge-list {male_gen_filenames} {self.thrflag} {filtering_flags} --make-bed --out {sub}_hwe',\n",
    "                        f'{funcName}_mergeSexes')\n",
    "        \n",
    "        self.genotypes_subset = f'{sub}_hwe'\n",
    "        \n",
    "    def generateGRM(self, autosome_list: list = list(range(1,21)), print_call: bool = False,\n",
    "                    extra_chrs: list = ['xchr']):\n",
    "        \n",
    "        '''\n",
    "        generates the grms, one per chromosome and one with all the chromossomes\n",
    "        '''\n",
    "        \n",
    "        funcName = inspect.getframeinfo(inspect.currentframe()).function\n",
    "                \n",
    "        self.bashLog(f'{self.gtca} {self.thrflag} --bfile {self.genotypes_subset} --autosome-num 21 --autosome --make-grm-bin --out {self.autoGRM}',\n",
    "            funcName, print_call = print_call)\n",
    "        \n",
    "        if 'xchr' in extra_chrs:\n",
    "            self.bashLog(f'{self.gtca} {self.thrflag} --bfile {self.genotypes_subset} --make-grm-xchr --out {self.xGRM}',\n",
    "                        f'{funcName}_chrX', print_call = print_call)\n",
    "        \n",
    "        for c in tqdm(autosome_list):\n",
    "            self.bashLog(f'{self.gtca} {self.thrflag} --bfile {self.genotypes_subset} --chr {c} --make-grm-bin --out {self.path}grm/{c}chrGRM',\n",
    "                        f'{funcName}_chr{c}',  print_call = print_call)\n",
    "\n",
    "    def snpHeritability(self, print_call: bool = False):\n",
    "        h2table = pd.DataFrame()\n",
    "        for trait in tqdm(self.traits):\n",
    "            trait_file = f'{self.path}data/pheno/{trait}.txt'\n",
    "            out_file = f'{self.path}results/heritability/{trait}' \n",
    "            df.dropna(subset = ['rfid'])[['rfid', 'rfid', trait]].fillna('NA').astype(str).to_csv(trait_file, index = False, sep = ' ', header = None)\n",
    "            \n",
    "            self.bashLog(f'{self.gtca} --reml {self.thrflag} --pheno {trait_file} --grm {self.autoGRM} --out {out_file}',\n",
    "                        f'snpHeritability_{trait}', print_call = print_call)\n",
    "            \n",
    "            a = pd.read_csv(f'{out_file}.hsq', skipfooter=6, sep = '\\t',engine='python')\n",
    "            b = pd.read_csv(f'{out_file}.hsq', skiprows=6, sep = '\\t', header = None, index_col = 0).T.rename({1: trait})\n",
    "            newrow = pd.concat(\n",
    "                [a[['Source','Variance']].T[1:].rename({i:j for i,j in enumerate(a.Source)}, axis = 1).rename({'Variance': trait}),\n",
    "                b],axis =1 )\n",
    "            h2table= pd.concat([h2table,newrow])\n",
    "            \n",
    "        h2table.to_csv(f'{self.path}results/heritability/heritability.tsv', sep = '\\t')\n",
    "        return h2table\n",
    "        \n",
    "    def gwasPerChr(self, nchr: int = 21, print_call: bool = False, append_run_db: bool = True):\n",
    "        for trait, chrom in tqdm(list(itertools.product(self.traits, range(1,nchr+1)))):\n",
    "            if chrom == 21: chrom = 'x'\n",
    "            self.bashLog(f'{self.gtca} {self.thrflag} --pheno {self.path}data/pheno/{trait}.txt --bfile {self.genotypes_subset} \\\n",
    "                                       --grm {self.path}grm/AllchrGRM   \\\n",
    "                                       --chr {chrom} \\\n",
    "                                       --mlma-subtract-grm {self.path}grm/{chrom}chrGRM  \\\n",
    "                                       --mlma --out {self.path}results/gwas/gwas_{chrom}_{trait}',\n",
    "                        f'GWAS_{chrom}_{trait}', print_call = print_call)\n",
    "    \n",
    "    def GWAS(self, subtract_grm: bool = False, loco: bool = True , print_call: bool = False):\n",
    "        #grm_flag = f'--grm {self.path}grm/AllchrGRM --mlma-subtract-grm {self.path}grm/AllchrGRM' if subtract_grm else ''\n",
    "        grm_name = 'sub_grm' if subtract_grm else 'with_grm'\n",
    "        loco_flag = '-loco' if loco else ''\n",
    "        for trait in tqdm(self.traits):\n",
    "            self.bashLog(f'{self.gtca} {self.thrflag} --pheno {self.path}data/pheno/{trait}.txt --bfile {self.genotypes_subset}\\\n",
    "                                       --mlma{loco_flag} --out {self.path}results/loco/{trait}',\n",
    "                        f'GWAS_{grm_name}_{loco_flag[1:]}_{trait}',  print_call = print_call)\n",
    "            \n",
    "    \n",
    "    \n",
    "    def addGWASresultsToDb(self, researcher, round_version, gwas_version,filenames: list = [], pval_thresh = 1e-3, **kwards):\n",
    "        \n",
    "        \n",
    "        if type(filenames) == str:\n",
    "            filenames = [filenames]\n",
    "            \n",
    "        elif len(filenames) == 0 :\n",
    "            filenames = sorted(glob(f'{self.path}results/loco/*.mlma'))\n",
    "        all_new = []\n",
    "        for file in tqdm(filenames):\n",
    "            trait = re.findall('/([^/]+).mlma', file)[0].replace('.loco', '')\n",
    "            new_info = pd.read_csv(file, sep = '\\t').assign(**kwards).assign(uploadeddate = datetime.today().strftime('%Y-%m-%d'),\n",
    "                                                                               researcher = researcher,\n",
    "                                                                               project = self.project_name,\n",
    "                                                                               trait = trait,\n",
    "                                                                               filename = file,\n",
    "                                                                               pval_threshold = pval_thresh,\n",
    "                                                                               genotypes_file = self.all_genotypes,\n",
    "                                                                               round_version = round_version,\n",
    "                                                                               gwas_version = gwas_version)\n",
    "\n",
    "            ###### add phewas version\n",
    "            new_info['n_snps'] =  new_info.shape[0]\n",
    "            # Subset by p-val 10e-2\n",
    "            new_info = new_info.query(f'p < {pval_thresh}')\n",
    "            all_new += [new_info]\n",
    "            # Add more to subset\n",
    "            # Test different keep values\n",
    "        all_new = pd.concat(all_new)   \n",
    "        \n",
    "        try:\n",
    "            alldata = pd.concat([all_new, pd.read_parquet(self.phewas_db)])\n",
    "        except:\n",
    "            print(f\"Could not open phewas database in file: {self.phewas_db}, rebuilding db with only this project\")\n",
    "            alldata = all_new\n",
    "        \n",
    "        alldata.drop_duplicates(subset = ['researcher', 'project', 'round_version', 'trait', 'SNP', 'uploadeddate'], \n",
    "                                keep='first').to_parquet(self.phewas_db, index = False, compression='gzip')\n",
    "        \n",
    "        ###scp this to tscc\n",
    "        \n",
    "        \n",
    "        \n",
    "    def callQTLs(self, threshold: float = 5.3, window: int = 1e6, subterm: int = 2,\n",
    "                 ldwin = 1e6, ldkb = 11000, ldr2 = .4, qtl_dist = 2*1e6, nchr: int = 21, NonStrictSearchDir = ''):\n",
    "        thresh = 10**(-threshold)\n",
    "        chr_list = ['x' if x == 21 else x for x in range(1,nchr+1)]\n",
    "        \n",
    "        if not NonStrictSearchDir:\n",
    "            topSNPs = pd.DataFrame()\n",
    "            for t, c in tqdm(list(itertools.product(self.traits, chr_list))):\n",
    "                filename = f'{self.path}results/gwas/gwas_{c}_{t}.mlma'\n",
    "                try:\n",
    "                    topSNPs = pd.concat([topSNPs, pd.read_csv(filename, sep = '\\t').query(f'p < {thresh}').assign(trait=t)])\n",
    "                except:\n",
    "                    print(f\"didn't open {filename}, does it exist?\")\n",
    "        else:\n",
    "            topSNPs = pd.concat([pd.read_csv(filename, sep = '\\t').query(f'p < {thresh}').assign(trait=re.findall('/([^/]*).mlma', \n",
    "                                                                                                                  filename)[0].replace('.loco', '')) for\n",
    "                                 filename in tqdm(glob(f\"{NonStrictSearchDir}/*.mlma\"))])\n",
    "\n",
    "        out = pd.DataFrame()\n",
    "\n",
    "        for (t, c), df in tqdm(topSNPs.groupby(['trait','Chr'])):\n",
    "            df = df.set_index('bp')\n",
    "            df.p = -np.log10(df.p)\n",
    "\n",
    "            while df.query('p > @threshold').shape[0]:\n",
    "                idx = df.p.idxmax()\n",
    "                maxp = df.loc[idx]\n",
    "                correlated_snps = df.loc[idx- window//2: idx + window//2].query('p > @maxp.p - @subterm')\n",
    "                qtl = True if correlated_snps.shape[0] > 2 else False\n",
    "\n",
    "                out = pd.concat([out,\n",
    "                                 maxp.to_frame().T.assign(QTL= qtl)],\n",
    "                                 axis = 0)\n",
    "\n",
    "                ldfilename = f'{self.path}temp/r2/temp_qtl_n_{t}'\n",
    "                self.bashLog(f'plink --bfile {self.genotypes_subset} --chr {c}  --ld-snp {maxp.SNP} \\\n",
    "                                     --ld-window {ldwin} {self.thrflag} \\\n",
    "                                     --nonfounders --r2  \\\n",
    "                                     --ld-window-r2 {ldr2} --out {ldfilename}',\n",
    "                             f'qlt_{t}', False )#--ld_window_kb {ldkb}\n",
    "\n",
    "                try: \n",
    "                    ldSNPS = pd.read_csv(f'{ldfilename}.ld', sep = r'\\s+').SNP_B.to_list() + [maxp.SNP]\n",
    "                    df = df.query('~(@idx - @qtl_dist//2 < index < @idx + @qtl_dist//2) and (SNP not in @ldSNPS)')\n",
    "                except:\n",
    "                    ldSNPS = [maxp.SNP]\n",
    "                    df = df.query('(SNP not in @ldSNPS)') ##### should this be different than the one above?\n",
    "                #if sum(cnt.values()) % 10 == 0: print(cnt)\n",
    "                            \n",
    "\n",
    "        out =  out.reset_index().rename({'index': 'bp'}, axis = 1).sort_values('trait')#.assign(project = self.project_name)\n",
    "        self.allqtlspath = f'{self.path}/results/qtls/allQTLS.csv'\n",
    "        out.to_csv(self.allqtlspath , index = False)\n",
    "        return out.set_index('SNP')   \n",
    "    \n",
    "    def phewas(self, qtltable, ld_window = int(3e6), pval_threshold = 1e-5, nreturn: int =1,r2_threshold = .8,\n",
    "              annotate = True, annotate_genome = 'rat6'):\n",
    "        '''\n",
    "        this function requires a QTL dataframe which is returned by callQTLs and a phewas database file\n",
    "        we will\n",
    "        \n",
    "        '''\n",
    "                \n",
    "        db_vals = pd.read_parquet(self.phewas_db, compression='gzip').query(f'p < {pval_threshold} and project != \"{self.project_name}\"')        \n",
    "        \n",
    "        table_exact_match = db_vals.merge(qtltable.reset_index(), on = 'SNP', how = 'inner', suffixes = ('_phewasdb', '_QTL'))\n",
    "        self.phewas_exact_match_path = f'{self.path}results/phewas/table_exact_match.csv'\n",
    "        table_exact_match.to_csv(self.phewas_exact_match_path )\n",
    "        #pd.concat([qtltable, db_vals] ,join = 'inner', axis = 1)\n",
    "        \n",
    "        nearby_snps = pd.concat([\n",
    "             self.plink(bfile = self.genotypes_subset, chr = row.Chr, ld_snp = row.name,\n",
    "               ld_window = ld_window, thread_num = 12, nonfounders = '', r2 = '')\\\n",
    "              .query(f'R2 > {r2_threshold}')\\\n",
    "              .drop(['CHR_A', 'BP_A', 'CHR_B'], axis = 1)\\\n",
    "              .rename({'SNP_A': 'SNP', 'SNP_B': 'NearbySNP', 'BP_B': 'NearbyBP'}, axis = 1)\\\n",
    "              .assign(**row.to_dict())\\\n",
    "              .set_index('SNP')\n",
    "              for  _, row in tqdm(list(qtltable.iterrows())) ])\n",
    "        \n",
    "        table_window_match = db_vals.merge(nearby_snps.reset_index(), left_on= 'SNP', \n",
    "                                                         right_on='NearbySNP', how = 'inner', suffixes = ('_phewasdb', '_QTL'))\n",
    "        \n",
    "        \n",
    "        self.phewas_window_r2 = f'{self.path}results/phewas/table_window_match_{ld_window}_{pval_threshold}_{r2_threshold}.csv'\n",
    "        \n",
    "        if table_window_match.shape[0] == 0:\n",
    "            print('No QTL window matches')\n",
    "            pd.DataFrame().to_csv(self.phewas_window_r2, index = False)\n",
    "            return -1\n",
    "            \n",
    "        \n",
    "        out = table_window_match.groupby([ 'SNP_QTL','project', 'trait_phewasdb'])\\\n",
    "                                .apply(lambda df : df[df.uploadeddate == df.uploadeddate.max()]\\\n",
    "                                                   .nsmallest(n = nreturn, columns = 'p_phewasdb'))\\\n",
    "                                                   .reset_index(drop = True)##.drop('index', axis = 1)\n",
    "        \n",
    "        out.to_csv(self.phewas_window_r2, index = False)\n",
    "        if annotate:\n",
    "            out = self.annotate(out.rename({'A1_phewasdb':'A1', 'A2_phewasdb': 'A2',\n",
    "                                            'Chr_phewasdb':'Chr', 'bp_phewasdb':'bp'}, axis = 1), annotate_genome, 'NearbySNP', save = False)\n",
    "        return out\n",
    "        \n",
    "\n",
    "        \n",
    "        \n",
    "    def eQTL(self, qtltable, pval_thresh: float = 1e-4, r2_thresh: float = .8, nreturn: int =1, ld_window: int = 3e6,\n",
    "            tissue_list: list = ['Adipose', 'BLA','Brain','Eye','IL','LHb','Liver','NAcc','NAcc2','OFC','PL','PL2'],\n",
    "            annotate = True, annotate_genome = 'rat6'):\n",
    "        '''\n",
    "        this function \n",
    "        \n",
    "        '''\n",
    "    \n",
    "        out = []\n",
    "        for tissue in tqdm(tissue_list,  position=0, desc=\"tissue\", leave=True):\n",
    "\n",
    "            tempdf = pd.read_csv(f'https://ratgtex.org/data/eqtl/{tissue}.cis_qtl_signif.txt.gz', sep = '\\t').assign(tissue = tissue)\\\n",
    "                                                                                                             .rename({'variant_id': 'SNP'}, axis = 1)\n",
    "            out += [pd.concat([\n",
    "                   self.plink(bfile = self.genotypes_subset, chr = row.Chr, ld_snp = row.name,\\\n",
    "                   ld_window = ld_window, thread_num = 12, nonfounders = '', r2 = '')\\\n",
    "                  .drop(['CHR_A', 'BP_A', 'CHR_B'], axis = 1)\\\n",
    "                  .rename({'SNP_A': 'SNP', 'SNP_B': 'NearbySNP', 'BP_B': 'NearbyBP'}, axis = 1)\\\n",
    "                  .assign(**row.to_dict())\\\n",
    "                  .merge(tempdf, right_on= 'SNP',  left_on='NearbySNP', how = 'inner', suffixes = ('_QTL', '_eqtldb'))\\\n",
    "                  .query(f'R2 > {r2_thresh} and pval_nominal < {pval_thresh}')\\\n",
    "                  .nsmallest(nreturn, 'pval_nominal')\n",
    "                  for  _, row in qtltable.iterrows() ])]\n",
    "\n",
    "        out = pd.concat(out).reset_index(drop=True)\n",
    "        if annotate:\n",
    "            out = self.annotate(out, annotate_genome, 'NearbySNP', save = False)\n",
    "        self.eqtl_path = f'{self.path}results/eqtl/eqtl.csv'\n",
    "        out.to_csv(self.eqtl_path, index= False)\n",
    "        \n",
    "        return out\n",
    "\n",
    "    def annotate(self, qtltable, genome: str = 'rat6', snpcol: str = 'SNP', save: bool = True):\n",
    "        \n",
    "        '''\n",
    "        we will use the snpeff to query the annotations for the QTL results and for the phewas results\n",
    "        '''\n",
    "        d = {'rat6': 'Rnor_6.0.99', 'rat7':'mRatBN7.2.105'}[genome]\n",
    "        #bash('java -jar snpEff/snpEff.jar download -v Rnor_6.0.99')\n",
    "        #bash('java -jar snpEff/snpEff.jar download -v mRatBN7.2.105')    \n",
    "        \n",
    "        temp  = qtltable.reset_index()\\\n",
    "                        .loc[:,[ 'Chr', 'bp', snpcol, 'A1', 'A2']]\\\n",
    "                        .assign(QUAL = 40, FILTER = 'PASS' ,INFO = '', FORMAT = 'GT:GQ:DP:HQ')\n",
    "        temp.columns = [\"##CHROM\",\"POS\",\"ID\",\"REF\",\"ALT\", 'QUAL', 'FILTER', 'INFO', 'FORMAT']\n",
    "        temp['##CHROM'] = 'chr'+ temp['##CHROM'].astype(str)\n",
    "        vcfGeneratorExitools.pandas2vcf(temp, f'{self.path}temp/test.vcf', metadata='')\n",
    "        a = bash(f'java -Xmx8g -jar snpEff/snpEff.jar {d} -no-intergenic -no-intron -noStats {self.path}temp/test.vcf', 'snpefftest' )\n",
    "        res =pd.read_csv(StringIO('\\n'.join(a)),  comment='#',  delim_whitespace=True,  header=None, names = temp.columns,  dtype=str)  \n",
    "        ann = res['INFO'].str.replace('ANN=', '').str.split('|',expand=True)\n",
    "        ann.columns = ['alt_temp', 'annotation', 'putative_impact', 'gene', 'geneid', 'featuretype', 'featureid', 'transcriptbiotype',\n",
    "                      'rank', 'HGVS.c', 'HGVS.p', 'cDNA_position|cDNA_len', 'CDS_position|CDS_len', 'Protein_position|Protein_len',\n",
    "                      'distancetofeature', 'errors']\n",
    "        ann.index = qtltable.index\n",
    "        out = pd.concat([qtltable, ann], axis = 1).replace('', np.nan).dropna(how = 'all', axis = 1).drop('alt_temp', axis = 1)\n",
    "        if save:\n",
    "            self.annotatedtablepath = f'{self.path}results/qtls/finalqtl.csv'\n",
    "            out.to_csv(self.annotatedtablepath, index= False)\n",
    "        \n",
    "        return out \n",
    "                \n",
    "\n",
    "    def print_watermark():\n",
    "        pass\n",
    "    \n",
    "\n",
    "            \n",
    "            \n",
    "            \n",
    "            \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "f97a961a-3220-4f61-9221-f1899ce98d2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('~/Documents/GitHub/sanchest/hsrats_round9_1/Normalized_filtered_tom_jhou_U01_lowercase.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b1d3cc3d-3dba-41ef-8142-d05e46dd2a00",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas = gwas_pipe(path = 'test/',\n",
    "                 all_genotypes = 'round9_1.vcf.gz',\n",
    "                 data = df,\n",
    "                 project_name = 'tj',\n",
    "                 traits =  df.loc[:, 'locomotor1':].columns.tolist(),\n",
    "                 threads=12,\n",
    "                )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "35b77cfd-a25a-458e-91f2-6955e06c6b13",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas.make_dir_structure()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0819328-5f48-44d6-911b-2f06f7c4965b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%time gwas.SubsampleMissMafHweFilter()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5fd0062-10ab-48d2-96ab-04a7eff11248",
   "metadata": {},
   "outputs": [],
   "source": [
    "%time gwas.generateGRM()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b915e3cf-25c0-4708-b438-93c170004f49",
   "metadata": {},
   "outputs": [],
   "source": [
    "%time gwas.snpHeritability()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bbc73d5-bf09-4b83-94d6-d01ac4fa051a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%time gwas.gwasPerChr()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f91d8d41-5974-44ef-8eef-5faafbdb407c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%time gwas.GWAS()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abb28363-08d5-4a05-b6b8-c586eb932fb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%time gwas.addGWASresultsToDb(researcher='tsanches', project='tj', round_version='9.1', gwas_version='0.0.1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c7841752-90c4-4580-8eeb-0276bd42919e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████████████████████████████████████████████| 15/15 [01:01<00:00,  4.07s/it]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████| 10/10 [00:11<00:00,  1.10s/it]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bp</th>\n",
       "      <th>Chr</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Freq</th>\n",
       "      <th>b</th>\n",
       "      <th>se</th>\n",
       "      <th>p</th>\n",
       "      <th>trait</th>\n",
       "      <th>QTL</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:143001817</th>\n",
       "      <td>143001817</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.069241</td>\n",
       "      <td>-0.48236</td>\n",
       "      <td>0.103227</td>\n",
       "      <td>5.527049</td>\n",
       "      <td>delaypunishmentratio</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr9:110362547</th>\n",
       "      <td>110362547</td>\n",
       "      <td>9</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.211093</td>\n",
       "      <td>0.351662</td>\n",
       "      <td>0.073998</td>\n",
       "      <td>5.696644</td>\n",
       "      <td>locomotor1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr9:111723693</th>\n",
       "      <td>111723693</td>\n",
       "      <td>9</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.199511</td>\n",
       "      <td>0.34787</td>\n",
       "      <td>0.074628</td>\n",
       "      <td>5.502892</td>\n",
       "      <td>locomotor1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr5:19455786</th>\n",
       "      <td>19455786</td>\n",
       "      <td>5</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>-0.251977</td>\n",
       "      <td>0.054721</td>\n",
       "      <td>5.384193</td>\n",
       "      <td>locomotor2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr11:20205102</th>\n",
       "      <td>20205102</td>\n",
       "      <td>11</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.410714</td>\n",
       "      <td>0.247502</td>\n",
       "      <td>0.053392</td>\n",
       "      <td>5.448529</td>\n",
       "      <td>locomotor2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr13:110174580</th>\n",
       "      <td>110174580</td>\n",
       "      <td>13</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.317942</td>\n",
       "      <td>-0.271674</td>\n",
       "      <td>0.059336</td>\n",
       "      <td>5.329553</td>\n",
       "      <td>progressiveratio</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr14:6165824</th>\n",
       "      <td>6165824</td>\n",
       "      <td>14</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.488064</td>\n",
       "      <td>0.245613</td>\n",
       "      <td>0.052523</td>\n",
       "      <td>5.534497</td>\n",
       "      <td>progressiveratio</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr2:72068315</th>\n",
       "      <td>72068315</td>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.147952</td>\n",
       "      <td>-0.370444</td>\n",
       "      <td>0.076975</td>\n",
       "      <td>5.826677</td>\n",
       "      <td>punishment</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr2:73124956</th>\n",
       "      <td>73124956</td>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.180387</td>\n",
       "      <td>-0.325543</td>\n",
       "      <td>0.068084</td>\n",
       "      <td>5.759393</td>\n",
       "      <td>punishment</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr2:75149524</th>\n",
       "      <td>75149524</td>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.182668</td>\n",
       "      <td>-0.318588</td>\n",
       "      <td>0.068973</td>\n",
       "      <td>5.41388</td>\n",
       "      <td>punishment</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3:37237803</th>\n",
       "      <td>37237803</td>\n",
       "      <td>3</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.35248</td>\n",
       "      <td>-0.26617</td>\n",
       "      <td>0.055719</td>\n",
       "      <td>5.749775</td>\n",
       "      <td>punishment</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr5:56029468</th>\n",
       "      <td>56029468</td>\n",
       "      <td>5</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.271726</td>\n",
       "      <td>0.3035</td>\n",
       "      <td>0.057375</td>\n",
       "      <td>6.911864</td>\n",
       "      <td>punishment</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr5:51018247</th>\n",
       "      <td>51018247</td>\n",
       "      <td>5</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.230469</td>\n",
       "      <td>0.331102</td>\n",
       "      <td>0.063413</td>\n",
       "      <td>6.750489</td>\n",
       "      <td>punishment</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr5:54308298</th>\n",
       "      <td>54308298</td>\n",
       "      <td>5</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.286156</td>\n",
       "      <td>0.284251</td>\n",
       "      <td>0.057153</td>\n",
       "      <td>6.182014</td>\n",
       "      <td>punishment</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr13:110174580</th>\n",
       "      <td>110174580</td>\n",
       "      <td>13</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.317942</td>\n",
       "      <td>0.273714</td>\n",
       "      <td>0.059547</td>\n",
       "      <td>5.367128</td>\n",
       "      <td>punishmenteffortratio</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        bp Chr A1 A2      Freq         b        se         p  \\\n",
       "SNP                                                                            \n",
       "chr1:143001817   143001817   1  T  C  0.069241  -0.48236  0.103227  5.527049   \n",
       "chr9:110362547   110362547   9  T  C  0.211093  0.351662  0.073998  5.696644   \n",
       "chr9:111723693   111723693   9  C  T  0.199511   0.34787  0.074628  5.502892   \n",
       "chr5:19455786     19455786   5  G  A  0.333333 -0.251977  0.054721  5.384193   \n",
       "chr11:20205102    20205102  11  C  T  0.410714  0.247502  0.053392  5.448529   \n",
       "chr13:110174580  110174580  13  T  C  0.317942 -0.271674  0.059336  5.329553   \n",
       "chr14:6165824      6165824  14  C  A  0.488064  0.245613  0.052523  5.534497   \n",
       "chr2:72068315     72068315   2  C  T  0.147952 -0.370444  0.076975  5.826677   \n",
       "chr2:73124956     73124956   2  G  T  0.180387 -0.325543  0.068084  5.759393   \n",
       "chr2:75149524     75149524   2  G  A  0.182668 -0.318588  0.068973   5.41388   \n",
       "chr3:37237803     37237803   3  G  A   0.35248  -0.26617  0.055719  5.749775   \n",
       "chr5:56029468     56029468   5  C  T  0.271726    0.3035  0.057375  6.911864   \n",
       "chr5:51018247     51018247   5  A  T  0.230469  0.331102  0.063413  6.750489   \n",
       "chr5:54308298     54308298   5  T  C  0.286156  0.284251  0.057153  6.182014   \n",
       "chr13:110174580  110174580  13  T  C  0.317942  0.273714  0.059547  5.367128   \n",
       "\n",
       "                                 trait    QTL  \n",
       "SNP                                            \n",
       "chr1:143001817    delaypunishmentratio  False  \n",
       "chr9:110362547              locomotor1   True  \n",
       "chr9:111723693              locomotor1   True  \n",
       "chr5:19455786               locomotor2   True  \n",
       "chr11:20205102              locomotor2   True  \n",
       "chr13:110174580       progressiveratio  False  \n",
       "chr14:6165824         progressiveratio   True  \n",
       "chr2:72068315               punishment   True  \n",
       "chr2:73124956               punishment   True  \n",
       "chr2:75149524               punishment   True  \n",
       "chr3:37237803               punishment   True  \n",
       "chr5:56029468               punishment   True  \n",
       "chr5:51018247               punishment   True  \n",
       "chr5:54308298               punishment   True  \n",
       "chr13:110174580  punishmenteffortratio  False  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#%time gwas.callQTLs()\n",
    "ot = gwas.callQTLs(NonStrictSearchDir = 'test/results/loco/')\n",
    "ot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54966ac1-8af7-4ee5-80b1-f69d32332ceb",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas.annotate(ot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df942811-3960-47e5-aec3-135b375b2939",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas.eQTL(ot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "11941328-c7c9-4558-ae77-83a58070691b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████████████████████████████████████████████| 15/15 [00:07<00:00,  1.99it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "java -Xmx8g -jar snpEff/snpEff.jar Rnor_6.0.99 -no-intergenic -no-intron -noStats test/temp/test.vcf\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>SNP_phewasdb</th>\n",
       "      <th>bp</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Freq_phewasdb</th>\n",
       "      <th>b_phewasdb</th>\n",
       "      <th>se_phewasdb</th>\n",
       "      <th>p_phewasdb</th>\n",
       "      <th>uploadeddate</th>\n",
       "      <th>...</th>\n",
       "      <th>QTL</th>\n",
       "      <th>annotation</th>\n",
       "      <th>putative_impact</th>\n",
       "      <th>gene</th>\n",
       "      <th>geneid</th>\n",
       "      <th>featuretype</th>\n",
       "      <th>featureid</th>\n",
       "      <th>transcriptbiotype</th>\n",
       "      <th>HGVS.c</th>\n",
       "      <th>distancetofeature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>chr11:20205102</td>\n",
       "      <td>20205102</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.410714</td>\n",
       "      <td>0.247502</td>\n",
       "      <td>0.053392</td>\n",
       "      <td>3.560170e-06</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "      <td>chr13:110174580</td>\n",
       "      <td>110174580</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.317942</td>\n",
       "      <td>-0.271674</td>\n",
       "      <td>0.059336</td>\n",
       "      <td>4.682170e-06</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13</td>\n",
       "      <td>chr13:110174580</td>\n",
       "      <td>110174580</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.317942</td>\n",
       "      <td>0.273714</td>\n",
       "      <td>0.059547</td>\n",
       "      <td>4.294100e-06</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14</td>\n",
       "      <td>chr14:6165824</td>\n",
       "      <td>6165824</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.488064</td>\n",
       "      <td>0.245613</td>\n",
       "      <td>0.052523</td>\n",
       "      <td>2.920810e-06</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1:143001817</td>\n",
       "      <td>143001817</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.069241</td>\n",
       "      <td>-0.482360</td>\n",
       "      <td>0.103227</td>\n",
       "      <td>2.971330e-06</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>chr2:72068315</td>\n",
       "      <td>72068315</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.147952</td>\n",
       "      <td>-0.370444</td>\n",
       "      <td>0.076975</td>\n",
       "      <td>1.490470e-06</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>chr2:73124956</td>\n",
       "      <td>73124956</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.180387</td>\n",
       "      <td>-0.325543</td>\n",
       "      <td>0.068084</td>\n",
       "      <td>1.740230e-06</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>chr2:75149524</td>\n",
       "      <td>75149524</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.182668</td>\n",
       "      <td>-0.318588</td>\n",
       "      <td>0.068973</td>\n",
       "      <td>3.855850e-06</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>U1</td>\n",
       "      <td>ENSRNOG00000052117</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSRNOT00000086797.1</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>n.-4593G&gt;A</td>\n",
       "      <td>4593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3</td>\n",
       "      <td>chr3:37237803</td>\n",
       "      <td>37237803</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.352480</td>\n",
       "      <td>-0.266170</td>\n",
       "      <td>0.055719</td>\n",
       "      <td>1.779200e-06</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>5</td>\n",
       "      <td>chr5:19455786</td>\n",
       "      <td>19455786</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>-0.251977</td>\n",
       "      <td>0.054721</td>\n",
       "      <td>4.128640e-06</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>5</td>\n",
       "      <td>chr5:51018247</td>\n",
       "      <td>51018247</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.230469</td>\n",
       "      <td>0.331102</td>\n",
       "      <td>0.063413</td>\n",
       "      <td>1.776280e-07</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>5</td>\n",
       "      <td>chr5:54308298</td>\n",
       "      <td>54308298</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.286156</td>\n",
       "      <td>0.284251</td>\n",
       "      <td>0.057153</td>\n",
       "      <td>6.576360e-07</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>5</td>\n",
       "      <td>chr5:56029468</td>\n",
       "      <td>56029468</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.271726</td>\n",
       "      <td>0.303500</td>\n",
       "      <td>0.057375</td>\n",
       "      <td>1.225000e-07</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>9</td>\n",
       "      <td>chr9:110362547</td>\n",
       "      <td>110362547</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.211093</td>\n",
       "      <td>0.351662</td>\n",
       "      <td>0.073998</td>\n",
       "      <td>2.010740e-06</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>9</td>\n",
       "      <td>chr9:111723693</td>\n",
       "      <td>111723693</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.199511</td>\n",
       "      <td>0.347870</td>\n",
       "      <td>0.074628</td>\n",
       "      <td>3.141290e-06</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chr     SNP_phewasdb         bp A1 A2  Freq_phewasdb  b_phewasdb  \\\n",
       "0    11   chr11:20205102   20205102  C  T       0.410714    0.247502   \n",
       "1    13  chr13:110174580  110174580  T  C       0.317942   -0.271674   \n",
       "2    13  chr13:110174580  110174580  T  C       0.317942    0.273714   \n",
       "3    14    chr14:6165824    6165824  C  A       0.488064    0.245613   \n",
       "4     1   chr1:143001817  143001817  T  C       0.069241   -0.482360   \n",
       "5     2    chr2:72068315   72068315  C  T       0.147952   -0.370444   \n",
       "6     2    chr2:73124956   73124956  G  T       0.180387   -0.325543   \n",
       "7     2    chr2:75149524   75149524  G  A       0.182668   -0.318588   \n",
       "8     3    chr3:37237803   37237803  G  A       0.352480   -0.266170   \n",
       "9     5    chr5:19455786   19455786  G  A       0.333333   -0.251977   \n",
       "10    5    chr5:51018247   51018247  A  T       0.230469    0.331102   \n",
       "11    5    chr5:54308298   54308298  T  C       0.286156    0.284251   \n",
       "12    5    chr5:56029468   56029468  C  T       0.271726    0.303500   \n",
       "13    9   chr9:110362547  110362547  T  C       0.211093    0.351662   \n",
       "14    9   chr9:111723693  111723693  C  T       0.199511    0.347870   \n",
       "\n",
       "    se_phewasdb    p_phewasdb uploadeddate  ...    QTL             annotation  \\\n",
       "0      0.053392  3.560170e-06   2023-01-12  ...   True                   None   \n",
       "1      0.059336  4.682170e-06   2023-01-12  ...  False                   None   \n",
       "2      0.059547  4.294100e-06   2023-01-12  ...  False                   None   \n",
       "3      0.052523  2.920810e-06   2023-01-12  ...   True                   None   \n",
       "4      0.103227  2.971330e-06   2023-01-12  ...  False                   None   \n",
       "5      0.076975  1.490470e-06   2023-01-12  ...   True                   None   \n",
       "6      0.068084  1.740230e-06   2023-01-12  ...   True                   None   \n",
       "7      0.068973  3.855850e-06   2023-01-12  ...   True  upstream_gene_variant   \n",
       "8      0.055719  1.779200e-06   2023-01-12  ...   True                   None   \n",
       "9      0.054721  4.128640e-06   2023-01-12  ...   True                   None   \n",
       "10     0.063413  1.776280e-07   2023-01-12  ...   True                   None   \n",
       "11     0.057153  6.576360e-07   2023-01-12  ...   True                   None   \n",
       "12     0.057375  1.225000e-07   2023-01-12  ...   True                   None   \n",
       "13     0.073998  2.010740e-06   2023-01-12  ...   True                   None   \n",
       "14     0.074628  3.141290e-06   2023-01-12  ...   True                   None   \n",
       "\n",
       "   putative_impact  gene              geneid featuretype  \\\n",
       "0             None  None                None        None   \n",
       "1             None  None                None        None   \n",
       "2             None  None                None        None   \n",
       "3             None  None                None        None   \n",
       "4             None  None                None        None   \n",
       "5             None  None                None        None   \n",
       "6             None  None                None        None   \n",
       "7         MODIFIER    U1  ENSRNOG00000052117  transcript   \n",
       "8             None  None                None        None   \n",
       "9             None  None                None        None   \n",
       "10            None  None                None        None   \n",
       "11            None  None                None        None   \n",
       "12            None  None                None        None   \n",
       "13            None  None                None        None   \n",
       "14            None  None                None        None   \n",
       "\n",
       "               featureid  transcriptbiotype      HGVS.c  distancetofeature  \n",
       "0                   None               None        None               None  \n",
       "1                   None               None        None               None  \n",
       "2                   None               None        None               None  \n",
       "3                   None               None        None               None  \n",
       "4                   None               None        None               None  \n",
       "5                   None               None        None               None  \n",
       "6                   None               None        None               None  \n",
       "7   ENSRNOT00000086797.1              snRNA  n.-4593G>A               4593  \n",
       "8                   None               None        None               None  \n",
       "9                   None               None        None               None  \n",
       "10                  None               None        None               None  \n",
       "11                  None               None        None               None  \n",
       "12                  None               None        None               None  \n",
       "13                  None               None        None               None  \n",
       "14                  None               None        None               None  \n",
       "\n",
       "[15 rows x 41 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas.phewas(ot, annotate=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d174539c-2987-4267-9721-1acd35b9898f",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp  = ot.reset_index().loc[:,[ 'Chr', 'bp', 'SNP', 'A1', 'A2']].assign(QUAL = 40, FILTER = 'PASS' ,INFO = '', FORMAT = 'GT:GQ:DP:HQ')\n",
    "temp.columns = [\"##CHROM\",\"POS\",\"ID\",\"REF\",\"ALT\", 'QUAL', 'FILTER', 'INFO', 'FORMAT']\n",
    "temp['##CHROM'] = 'chr'+ temp['##CHROM'].astype(str)\n",
    "vcfGeneratorExitools.pandas2vcf(temp, f'{gwas.path}temp/test.vcf', metadata='' )\n",
    "a = bash(f'java -Xmx8g -jar snpEff/snpEff.jar Rnor_6.0.99 -no-intergenic -no-intron -noStats {gwas.path}temp/test.vcf', 'snpefftest' )\n",
    "res =pd.read_csv(StringIO('\\n'.join(a)),  comment='#',  delim_whitespace=True,  header=None, names = temp.columns,  dtype=str)  \n",
    "ann = res['INFO'].str.replace('ANN=', '').str.split('|',expand=True)\n",
    "ann.columns = ['alt', 'annotation', 'putative_impact', 'gene', 'geneid', 'featuretype', 'featureid', 'transcriptbiotype',\n",
    "              'rank', 'HGVS.c', 'HGVS.p', 'cDNA_position|cDNA_len', 'CDS_position|CDS_len', 'Protein_position|Protein_len',\n",
    "              'distancetofeature', 'errors']\n",
    "ann.index = ot.index\n",
    "out = pd.concat([ot, ann], axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "df30fcb5-1370-417e-9de1-bad2bf625e9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5.011872336272724e-06"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1/10**5.3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "607a07fd-ad46-42be-b618-bf141ece144c",
   "metadata": {},
   "outputs": [],
   "source": [
    "db_vals = pd.read_parquet(gwas.phewas_db).query(f'p < {1e-5} and project != \"{gwas.project_name}\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fd8e223f-92f9-4d6b-91eb-48660f938e5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>SNP</th>\n",
       "      <th>bp</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Freq</th>\n",
       "      <th>b</th>\n",
       "      <th>se</th>\n",
       "      <th>p</th>\n",
       "      <th>uploadeddate</th>\n",
       "      <th>researcher</th>\n",
       "      <th>project</th>\n",
       "      <th>trait</th>\n",
       "      <th>filename</th>\n",
       "      <th>genotypes_file</th>\n",
       "      <th>round_version</th>\n",
       "      <th>gwas_version</th>\n",
       "      <th>n_snps</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7507695</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1:142947467</td>\n",
       "      <td>142947467</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.074440</td>\n",
       "      <td>-0.443599</td>\n",
       "      <td>0.100037</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>tsanches</td>\n",
       "      <td>tomjhou2</td>\n",
       "      <td>delaypunishmentratio</td>\n",
       "      <td>test/results/loco/delaypunishmentratio.loco.mlma</td>\n",
       "      <td>round9_1.vcf.gz</td>\n",
       "      <td>9.1</td>\n",
       "      <td>0.0.1</td>\n",
       "      <td>4598918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7507697</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1:142948896</td>\n",
       "      <td>142948896</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.074440</td>\n",
       "      <td>-0.443599</td>\n",
       "      <td>0.100037</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>tsanches</td>\n",
       "      <td>tomjhou2</td>\n",
       "      <td>delaypunishmentratio</td>\n",
       "      <td>test/results/loco/delaypunishmentratio.loco.mlma</td>\n",
       "      <td>round9_1.vcf.gz</td>\n",
       "      <td>9.1</td>\n",
       "      <td>0.0.1</td>\n",
       "      <td>4598918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7507698</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1:142949913</td>\n",
       "      <td>142949913</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.074440</td>\n",
       "      <td>-0.443599</td>\n",
       "      <td>0.100037</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>tsanches</td>\n",
       "      <td>tomjhou2</td>\n",
       "      <td>delaypunishmentratio</td>\n",
       "      <td>test/results/loco/delaypunishmentratio.loco.mlma</td>\n",
       "      <td>round9_1.vcf.gz</td>\n",
       "      <td>9.1</td>\n",
       "      <td>0.0.1</td>\n",
       "      <td>4598918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7507700</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1:142952269</td>\n",
       "      <td>142952269</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.074440</td>\n",
       "      <td>-0.443599</td>\n",
       "      <td>0.100037</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>tsanches</td>\n",
       "      <td>tomjhou2</td>\n",
       "      <td>delaypunishmentratio</td>\n",
       "      <td>test/results/loco/delaypunishmentratio.loco.mlma</td>\n",
       "      <td>round9_1.vcf.gz</td>\n",
       "      <td>9.1</td>\n",
       "      <td>0.0.1</td>\n",
       "      <td>4598918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7507701</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1:142953106</td>\n",
       "      <td>142953106</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.074440</td>\n",
       "      <td>-0.443599</td>\n",
       "      <td>0.100037</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>tsanches</td>\n",
       "      <td>tomjhou2</td>\n",
       "      <td>delaypunishmentratio</td>\n",
       "      <td>test/results/loco/delaypunishmentratio.loco.mlma</td>\n",
       "      <td>round9_1.vcf.gz</td>\n",
       "      <td>9.1</td>\n",
       "      <td>0.0.1</td>\n",
       "      <td>4598918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12030975</th>\n",
       "      <td>13</td>\n",
       "      <td>chr13:110188376</td>\n",
       "      <td>110188376</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.319317</td>\n",
       "      <td>0.264789</td>\n",
       "      <td>0.059338</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>tsanches</td>\n",
       "      <td>tomjhou2</td>\n",
       "      <td>punishmenteffortratio</td>\n",
       "      <td>test/results/loco/punishmenteffortratio.loco.mlma</td>\n",
       "      <td>round9_1.vcf.gz</td>\n",
       "      <td>9.1</td>\n",
       "      <td>0.0.1</td>\n",
       "      <td>4598918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12030976</th>\n",
       "      <td>13</td>\n",
       "      <td>chr13:110188437</td>\n",
       "      <td>110188437</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.319317</td>\n",
       "      <td>0.264789</td>\n",
       "      <td>0.059338</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>tsanches</td>\n",
       "      <td>tomjhou2</td>\n",
       "      <td>punishmenteffortratio</td>\n",
       "      <td>test/results/loco/punishmenteffortratio.loco.mlma</td>\n",
       "      <td>round9_1.vcf.gz</td>\n",
       "      <td>9.1</td>\n",
       "      <td>0.0.1</td>\n",
       "      <td>4598918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12031518</th>\n",
       "      <td>13</td>\n",
       "      <td>chr13:110681227</td>\n",
       "      <td>110681227</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.323963</td>\n",
       "      <td>0.265641</td>\n",
       "      <td>0.059408</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>tsanches</td>\n",
       "      <td>tomjhou2</td>\n",
       "      <td>punishmenteffortratio</td>\n",
       "      <td>test/results/loco/punishmenteffortratio.loco.mlma</td>\n",
       "      <td>round9_1.vcf.gz</td>\n",
       "      <td>9.1</td>\n",
       "      <td>0.0.1</td>\n",
       "      <td>4598918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12031867</th>\n",
       "      <td>13</td>\n",
       "      <td>chr13:110904823</td>\n",
       "      <td>110904823</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.313673</td>\n",
       "      <td>0.269657</td>\n",
       "      <td>0.060125</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>tsanches</td>\n",
       "      <td>tomjhou2</td>\n",
       "      <td>punishmenteffortratio</td>\n",
       "      <td>test/results/loco/punishmenteffortratio.loco.mlma</td>\n",
       "      <td>round9_1.vcf.gz</td>\n",
       "      <td>9.1</td>\n",
       "      <td>0.0.1</td>\n",
       "      <td>4598918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12031868</th>\n",
       "      <td>13</td>\n",
       "      <td>chr13:110904828</td>\n",
       "      <td>110904828</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.313423</td>\n",
       "      <td>0.265784</td>\n",
       "      <td>0.060146</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>tsanches</td>\n",
       "      <td>tomjhou2</td>\n",
       "      <td>punishmenteffortratio</td>\n",
       "      <td>test/results/loco/punishmenteffortratio.loco.mlma</td>\n",
       "      <td>round9_1.vcf.gz</td>\n",
       "      <td>9.1</td>\n",
       "      <td>0.0.1</td>\n",
       "      <td>4598918</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5614 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Chr              SNP         bp A1 A2      Freq         b        se  \\\n",
       "7507695     1   chr1:142947467  142947467  A  G  0.074440 -0.443599  0.100037   \n",
       "7507697     1   chr1:142948896  142948896  T  C  0.074440 -0.443599  0.100037   \n",
       "7507698     1   chr1:142949913  142949913  T  C  0.074440 -0.443599  0.100037   \n",
       "7507700     1   chr1:142952269  142952269  T  C  0.074440 -0.443599  0.100037   \n",
       "7507701     1   chr1:142953106  142953106  T  C  0.074440 -0.443599  0.100037   \n",
       "...       ...              ...        ... .. ..       ...       ...       ...   \n",
       "12030975   13  chr13:110188376  110188376  A  G  0.319317  0.264789  0.059338   \n",
       "12030976   13  chr13:110188437  110188437  T  C  0.319317  0.264789  0.059338   \n",
       "12031518   13  chr13:110681227  110681227  T  G  0.323963  0.265641  0.059408   \n",
       "12031867   13  chr13:110904823  110904823  T  C  0.313673  0.269657  0.060125   \n",
       "12031868   13  chr13:110904828  110904828  T  G  0.313423  0.265784  0.060146   \n",
       "\n",
       "                 p uploadeddate researcher   project                  trait  \\\n",
       "7507695   0.000009   2023-01-12   tsanches  tomjhou2   delaypunishmentratio   \n",
       "7507697   0.000009   2023-01-12   tsanches  tomjhou2   delaypunishmentratio   \n",
       "7507698   0.000009   2023-01-12   tsanches  tomjhou2   delaypunishmentratio   \n",
       "7507700   0.000009   2023-01-12   tsanches  tomjhou2   delaypunishmentratio   \n",
       "7507701   0.000009   2023-01-12   tsanches  tomjhou2   delaypunishmentratio   \n",
       "...            ...          ...        ...       ...                    ...   \n",
       "12030975  0.000008   2023-01-12   tsanches  tomjhou2  punishmenteffortratio   \n",
       "12030976  0.000008   2023-01-12   tsanches  tomjhou2  punishmenteffortratio   \n",
       "12031518  0.000008   2023-01-12   tsanches  tomjhou2  punishmenteffortratio   \n",
       "12031867  0.000007   2023-01-12   tsanches  tomjhou2  punishmenteffortratio   \n",
       "12031868  0.000010   2023-01-12   tsanches  tomjhou2  punishmenteffortratio   \n",
       "\n",
       "                                                   filename   genotypes_file  \\\n",
       "7507695    test/results/loco/delaypunishmentratio.loco.mlma  round9_1.vcf.gz   \n",
       "7507697    test/results/loco/delaypunishmentratio.loco.mlma  round9_1.vcf.gz   \n",
       "7507698    test/results/loco/delaypunishmentratio.loco.mlma  round9_1.vcf.gz   \n",
       "7507700    test/results/loco/delaypunishmentratio.loco.mlma  round9_1.vcf.gz   \n",
       "7507701    test/results/loco/delaypunishmentratio.loco.mlma  round9_1.vcf.gz   \n",
       "...                                                     ...              ...   \n",
       "12030975  test/results/loco/punishmenteffortratio.loco.mlma  round9_1.vcf.gz   \n",
       "12030976  test/results/loco/punishmenteffortratio.loco.mlma  round9_1.vcf.gz   \n",
       "12031518  test/results/loco/punishmenteffortratio.loco.mlma  round9_1.vcf.gz   \n",
       "12031867  test/results/loco/punishmenteffortratio.loco.mlma  round9_1.vcf.gz   \n",
       "12031868  test/results/loco/punishmenteffortratio.loco.mlma  round9_1.vcf.gz   \n",
       "\n",
       "         round_version gwas_version   n_snps  \n",
       "7507695            9.1        0.0.1  4598918  \n",
       "7507697            9.1        0.0.1  4598918  \n",
       "7507698            9.1        0.0.1  4598918  \n",
       "7507700            9.1        0.0.1  4598918  \n",
       "7507701            9.1        0.0.1  4598918  \n",
       "...                ...          ...      ...  \n",
       "12030975           9.1        0.0.1  4598918  \n",
       "12030976           9.1        0.0.1  4598918  \n",
       "12031518           9.1        0.0.1  4598918  \n",
       "12031867           9.1        0.0.1  4598918  \n",
       "12031868           9.1        0.0.1  4598918  \n",
       "\n",
       "[5614 rows x 18 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "        nearby_snps = pd.concat([\n",
    "             self.plink(bfile = self.genotypes_subset, chr = row.Chr, ld_snp = row.name,\n",
    "               ld_window = ld_window, thread_num = 12, nonfounders = '', r2 = '')\\\n",
    "              .query(f'R2 > {r2_threshold}')\\\n",
    "              .drop(['CHR_A', 'BP_A', 'CHR_B'], axis = 1)\\\n",
    "              .rename({'SNP_A': 'SNP', 'SNP_B': 'NearbySNP', 'BP_B': 'NearbyBP'}, axis = 1)\\\n",
    "              .assign(**row.to_dict())\\\n",
    "              .set_index('SNP')\n",
    "              for  _, row in tqdm(list(qtltable.iterrows())) ])\n",
    "        \n",
    "        table_window_match = db_vals.merge(nearby_snps.reset_index(), left_on= 'SNP', \n",
    "                                                         right_on='NearbySNP', how = 'inner', suffixes = ('_phewasdb', '_QTL'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d5a7633-d19f-4db6-87cf-41d3210d9460",
   "metadata": {},
   "outputs": [],
   "source": [
    "test.groupby('project').apply(lambda df: df.query(f'uploadeddate == \"{test.uploadeddate.max()}\"'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1288c931-323a-4cb7-b45d-a8fb7dcb8ca8",
   "metadata": {},
   "outputs": [],
   "source": [
    "test.nlargest(1, columns ='uploadeddate', keep = 'all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12b0856c-470e-416e-9f68-edadf4163c78",
   "metadata": {},
   "outputs": [],
   "source": [
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffea9be9-354d-4368-8816-bdc0cd5088ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0d3bd5d-c7b9-4d16-b85a-d0c46bbadb7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue_list = ['Adipose', 'BLA','Brain','Eye','IL','LHb','Liver','NAcc','NAcc2','OFC','PL','PL2' ]\n",
    "alleqtls = pd.concat([pd.read_csv(f'https://ratgtex.org/data/eqtl/{tissue}.cis_qtl_signif.txt.gz', sep = '\\t').assign(tissue = tissue)\n",
    "          for tissue in  tqdm(tissue_list)  ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caf1a24e-1415-4cac-b885-014a8a6b66fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "pval_thresh = 1e-4\n",
    "r2_thresh = .8\n",
    "nreturn =1 #int( 1e5)\n",
    "tissue_list = ['Adipose', 'BLA','Brain','Eye','IL','LHb','Liver','NAcc','NAcc2','OFC','PL','PL2' ]\n",
    "\n",
    "out = []\n",
    "for tissue in tqdm(tissue_list[:2],  position=0, desc=\"tissue\", leave=True):\n",
    "\n",
    "    tempdf = pd.read_csv(f'https://ratgtex.org/data/eqtl/{tissue}.cis_qtl_signif.txt.gz', sep = '\\t').assign(tissue = tissue)\\\n",
    "                                                                                                             .rename({'variant_id': 'SNP'}, axis = 1)\n",
    "    out += [pd.concat([\n",
    "         gwas.plink(bfile = gwas.genotypes_subset, chr = row.Chr, ld_snp = row.name,\\\n",
    "           ld_window = 1e6, thread_num = 12, nonfounders = '', r2 = '')\\\n",
    "          .drop(['CHR_A', 'BP_A', 'CHR_B'], axis = 1)\\\n",
    "          .rename({'SNP_A': 'SNP', 'SNP_B': 'NearbySNP', 'BP_B': 'NearbyBP'}, axis = 1)\\\n",
    "          .assign(**row.to_dict())\\\n",
    "          .merge(tempdf, right_on= 'SNP',  left_on='NearbySNP', how = 'inner', suffixes = ('_QTL', '_eqtldb'))\n",
    "          .query(f'R2 > {r2_thresh} and pval_nominal < {pval_thresh}')\\\n",
    "          .nsmallest(nreturn, 'pval_nominal')\n",
    "          for  _, row in tqdm(list(ot.iterrows()),  position=0, desc=\"qtl_row\", leave=True) ])]\n",
    "\n",
    "out = pd.concat(out).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56307c26-9e84-46a2-987b-cefe1dd2faaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb0d49b5-7b1e-4bfc-8afb-016b08f10dee",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (_, row), tissue in list(product(ot.iterrows(), tissue_list)):\n",
    "    display(row)\n",
    "    print(tissue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d0619d4-aa10-4c1a-bf99-c6972894fe53",
   "metadata": {},
   "outputs": [],
   "source": [
    "table_window_match"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d6a6821-13e2-4e04-a327-7895c4b41f6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas.bashLog(f'java -Xmx8g -jar snpEff.jar', \"tesi\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20c2b179-f8af-4eff-bc10-898dd9fa9af7",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "!find /home/bonnie/miniconda3/envs/dask -name \"snpEff.jar\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b944a71c-040d-439a-bc0e-7eb145efdff8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !conda install -c conda-forge snpeff -y\n",
    "!java -jar snpEff.jar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad4163e3-0967-42ef-bc9c-72a2699a570c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# java-jdk                  8.0.112                       1    bioconda\n",
    "# snpeff                    5.1                  hdfd78af_0    bioconda\n",
    "# !conda list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab7aef7f-f49a-4668-b70d-b32293e0823e",
   "metadata": {},
   "outputs": [],
   "source": [
    "!conda install -c bioconda java-jdk=8 -y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbcdf09f-9acd-4f04-9c90-60063f1b831c",
   "metadata": {},
   "outputs": [],
   "source": [
    "!conda install -c conda-forge openjdk=8 -y\n",
    "!conda install -c bioconda snpeff -y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c957f025-c180-47d9-a4a5-2e1f92d367cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "table_window_match = db_vals.reset_index().merge(nearby_snps.reset_index(), left_on= 'SNP', \n",
    "                                                         right_on='NearbySNP', how = 'inner', suffixes = ('_phewasdb', '_QTL'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9c38825-9125-45b3-b1f8-7577f35a26f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "table_window_match.R2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54bb7b9a-9307-4c1f-bacf-05c78baabe48",
   "metadata": {},
   "outputs": [],
   "source": [
    "db_vals = pd.read_parquet(gwas.phewas_db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3016019-52e7-49ca-803b-2287394a314f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc88196f-8d85-4554-9b37-4fd75d7a60fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "rg = gwas.plink(bfile = gwas.genotypes_subset, chr = testtest.Chr, ld_snp = testtest.name,\n",
    "      ld_window = 1e6, thread_num = 12, nonfounders = '', r2 = '')\\\n",
    "      .drop(['CHR_A', 'BP_A', 'CHR_B'], axis = 1)\\\n",
    "      .rename({'SNP_A': 'SNP', 'SNP_B': 'NearbySNP', 'BP_B': 'NearbyBP'}, axis = 1)\\\n",
    "      .assign(**testtest.to_dict())\\\n",
    "      .set_index('SNP')\n",
    "\n",
    "rg.query('R2 > .6')#['BP_B'].agg({\"max\":'max',\"min\": 'min'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92c64fd1-678d-42bb-8bdc-410a2fa5e4f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pull_data_from_db(project_name, savefiles = True):\n",
    "    out = {}\n",
    "    \n",
    "    connection = psycopg2.connect(database = \"PalmerLab_Datasets\",\n",
    "                              user = \"postgres\",\n",
    "                              password = \"palmerlab-amapostgres\",\n",
    "                              host = \"palmerlab-main-database-c2021-08-02.c6sgfwysomht.us-west-2.rds.amazonaws.com\",\n",
    "                              port = '5432')\n",
    "    cursor = connection.cursor()\n",
    "    \n",
    "    for key, val in {'wfu': f'{project_name}.wfu_master', \n",
    "                     'phenotypes': f'{project_name}.gwas_phenotypes', \n",
    "                     'tissue': f'sample_tracking.tissue' }.items():\n",
    "        try:\n",
    "            out[key] =  pd.read_sql(f\"SELECT * FROM {val}\", con = connection)\n",
    "            if savefiles:\n",
    "                os.makedirs(f'{project_name}', exist_ok = True)\n",
    "                out[key].to_csv(f'{project_name}/db_{val}.csv', index = False)\n",
    "        except:\n",
    "            print(f'could not find {project_name}/{val}')\n",
    "            out[key] = pd.DataFrame()\n",
    "\n",
    "    cursor.close()\n",
    "    connection.close()\n",
    "    return out\n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7013fe99-a3d3-4c17-afa4-cea9a38a0459",
   "metadata": {},
   "outputs": [],
   "source": [
    "phewas_projects = ['p50_david_dietz_2020', 'p50_jerry_richards_2014', 'u01_suzanne_mitchell',\n",
    "                   'p50_paul_meyer_2014', 'p50_paul_meyer_2020', 'r01_doug_adams',\n",
    "                  'u01_olivier_george_cocaine', 'u01_olivier_george_oxycodone', 'u01_peter_kalivas_italy',\n",
    "                  'u01_peter_kalivas_us', 'u01_tom_jhou', 'dean_baculum', 'p50_shelly_flagel_2014',\n",
    "                   'lionikas_2014']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "167174ca-770b-407a-a172-18ae6dc00ede",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in phewas_projects:\n",
    "    pull_data_from_db(i)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3b4920d-2d13-4773-b464-90c3196ceb2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "phewas_projects_avaliable = pd.DataFrame(sorted(glob('**/*.gwas_pheno*')), columns = ['phenotype_files'])\n",
    "phewas_projects_avaliable['project'] = phewas_projects_avaliable.phenotype_files.apply(lambda x: x.split('/')[-2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba4a5804-9275-4fbd-a318-166c51d633c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "phewas_projects_avaliable[['samplesize', 'phenotypesN']] = phewas_projects_avaliable.phenotype_files.apply(lambda x: list(pd.read_csv(x).shape)).tolist()\n",
    "phewas_projects_avaliable = phewas_projects_avaliable.assign(genotypes_version = str(9.1), \n",
    "                                                             researcher = 'sanchest', \n",
    "                                                             gwasrundate = datetime.today().strftime('%Y-%m-%d'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46b8cacf-0efe-4629-950e-60f9eb05bd81",
   "metadata": {},
   "outputs": [],
   "source": [
    "phewas_projects_avaliable.to_csv('phewas_info_db.csv',)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5830d873-ac0e-4d57-8706-6228020eda31",
   "metadata": {},
   "outputs": [],
   "source": [
    "phewas_projects_avaliable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "718b51ae-7ba6-475a-87eb-12ade8e94a3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "for project in phewas_projects_avaliable.iterrows():\n",
    "    gwas = \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d69db84-6b26-4d21-bc0e-7aeae001d8fd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5aedb31-afc4-43b7-9638-fe2907a0b5f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "phewas_projects_avaliable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7c74b81-a2f1-4049-87a4-eb5762233293",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_rattacca = pd.read_csv('~/Documents/GitHub/sanchest/rattaca/CompleteG01.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8909fb4b-2e7d-42ad-989c-1928ed9d6d67",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas = gwas_pipe(path = 'rattacca/',\n",
    "                 all_genotypes = 'round9_1.vcf.gz',\n",
    "                 data = df,\n",
    "                 traits =  df.loc[:, 'locomotor1':].columns.tolist(),\n",
    "                 threads=12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "feec2c32-61cb-43c5-89d2-cfb5751c985e",
   "metadata": {},
   "outputs": [],
   "source": [
    "rids with 3 zeros are gbs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8147d523-1b0c-469b-b6ae-a970a8c46205",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c10c7313-363a-4866-94c2-5e81200b6c75",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1260c854-b2d0-4cc6-9d41-5fe3e75bc243",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "########################## download main vcf file round 9.1\n",
    "#scp tsanches@tscc-login.sdsc.edu:/projects/ps-palmer/hs_rats/round9_1/Heterogenous-stock_n14780_10182022_QC_Sex_Het_pass_n13548.vcf.gz round9_1.vcf.gz\n",
    "#scp tsanches@tscc-login.sdsc.edu:/projects/ps-palmer/hs_rats/round9_1/Heterogenous-stock_n14780_10182022_QC_Sex_Het_pass_n13548.vcf.gz.tbi round9_1.vcf.gz.tbi\n",
    "#scp tsanches@tscc-login.sdsc.edu:/projects/ps-palmer/b3peng/gwas/p50_paul_meyer_2020/genotypes/dosages/dosages.txt round9_1_dosages.tsv\n",
    "#scp tsanches@tscc-login.sdsc.edu:/projects/ps-palmer/tsanches/reference_panel/HOMCONC_FINAL_CHROMPOS.tsv round10CHRMPOS.tsv\n",
    "#scp tsanches@tscc-login.sdsc.edu:/projects/ps-palmer/hs_rats/rattacafinalgenotypes/RattacaG01/RattacaG01_QC_Sex_Het_pass_n971.vcf.gz.tbi rattaca_genotypes.vcf.gz.tbi\n",
    "\n",
    "\n",
    "        #print('geno maf filtering')\n",
    "        #self.bashLog(f'plink --bfile {sub}_hwe --geno {geno} --maf {maf} --thread-num 12 --make-bed --out {sub}',\n",
    "        #                f'{funcName}_geno_maf')\n",
    "        \n",
    "        ### LD pruning \n",
    "        #self.bashLog(f'plink --bfile {sub} --indep-pairwise 50 5 0.999 --thread-num 12 --out {sub}_pruned_50_5_999',\n",
    "        #                f'{funcName}_findLDs')\n",
    "        \n",
    "        #self.bashLog(f'plink --bfile {sub}_hwe --geno {geno} --maf {maf} --thread-num 12 --make-bed --out {sub}_pruned',\n",
    "        #                f'{funcName}_prune')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:gpipe]",
   "language": "python",
   "name": "conda-env-gpipe-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
